### ISSUE TITLE
[fsdp2] maybe unreliable `set_unshard_in_backward(False)`

### ISSUE THREAD

**Issue by leonardo0lyj** [pytorch/pytorch#144800]
Hey Andrew @awgu,

As a big fan of FSDP2, I keep posting improvement 😄

This flag ([`set_unshard_in_backward(False)`](https://github.com/pytorch/pytorch/blob/aa57f0c6637d4377d2d86d377fdf41840498960a/torch/distributed/fsdp/_fully_shard/_fully_shard.py#L408)) is super helpful to skip `unshard()` in backward pass especially for unused backward pass of certain `nn.Module`. 

For valid example -- ZeRO3, these certain parameters are only used in forward, but not used in backward (skipped `unshard()`), so we keep those parameters in sharded state and keep gradient in None during backward pass, which saves allgather and reduce communication while being correct.

But in case of ZeRO2 or ZeRO++ (`reshard_after_forward=False or int`), this `set_unshard_in_backward(False)` results in misleading logic and potential bug.  For example, 
- these certain parameters are only used in forward, but not used in backward
- after forward, these parameters stay in unsharded state (ZeRO2) or [replicate, shard] state (ZeRO++)
- `set_unshard_in_backward(False)` skips `unshard` in pre-backward hook
- so in backward, these parameters stay in ZeRO2 and ZeRO++ state, which is misleading enough (i.e., by semantic, this flag `set_unshard_in_backward(False)` should mean that parameters stays in sharded state but in fact in ZeRO2 and ZeRO++ state)

As a suggestion, we can either ban ZeRO2 and ZeRO++ when using this flag`set_unshard_in_backward(False)` or figure out a more general way to support all ZeRO state.

How do you think 😁?

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

**Comment by awgu** [comment]
I think this is kind of hard. The `set_unshard_in_backward` is explicitly a flag to skip any unshard in backward. We could ban it from being used when `reshard_after_forward is not True`. It is probably better to keep the meaning of the flag simple: just do not try to unshard in backward.
